primitives/trie/src/cache/mod.rs

```diff
@@ -36,43 +36,285 @@
 
 use crate::{Error, NodeCodec};
 use hash_db::Hasher;
-use hashbrown::HashSet;
 use nohash_hasher::BuildNoHashHasher;
-use parking_lot::{Mutex, MutexGuard, RwLockReadGuard};
-use shared_cache::{SharedValueCache, ValueCacheKey};
+use parking_lot::{Mutex, MutexGuard};
+use schnellru::LruMap;
+use shared_cache::{ValueCacheKey, ValueCacheRef};
 use std::{
-	collections::{hash_map::Entry as MapEntry, HashMap},
-	sync::Arc,
+	collections::HashMap,
+	sync::{
+		atomic::{AtomicU64, Ordering},
+		Arc,
+	},
+	time::Duration,
 };
 use trie_db::{node::NodeOwned, CachedValue};
 
 mod shared_cache;
 
 pub use shared_cache::SharedTrieCache;
 
-use self::shared_cache::{SharedTrieCacheInner, ValueCacheKeyHash};
+use self::shared_cache::ValueCacheKeyHash;
 
 const LOG_TARGET: &str = "trie-cache";
 
-/// The size of the cache.
+/// The maximum amount of time we'll wait trying to acquire the shared cache lock
+/// when the local cache is dropped and synchronized with the share cache.
+///
+/// This is just a failsafe; normally this should never trigger.
+const SHARED_CACHE_WRITE_LOCK_TIMEOUT: Duration = Duration::from_millis(100);
+
+/// The maximum number of existing keys in the shared cache that a single local cache
+/// can promote to the front of the LRU cache in one go.
+///
+/// If we have a big shared cache and the local cache hits all of those keys we don't
+/// want to spend forever bumping all of them.
+const SHARED_NODE_CACHE_MAX_PROMOTED_KEYS: u32 = 1792;
+/// Same as [`SHARED_NODE_CACHE_MAX_PROMOTED_KEYS`].
+const SHARED_VALUE_CACHE_MAX_PROMOTED_KEYS: u32 = 1792;
+
+/// The maximum portion of the shared cache (in percent) that a single local
+/// cache can replace in one go.
+///
+/// We don't want a single local cache instance to have the ability to replace
+/// everything in the shared cache.
```

Comment:
They kinda serve different purposes.

The `MAX_REPLACE_PERCENT` is there so that a big local cache can't replace the whole of a small shared cache; the `MAX_PROMOTED_KEYS` is there so that we don't spend too much time shuffling entries around in a big shared cache.

Reply:
:+1: I should read the comments more carefully.